File: server/app/service/auth_service.py
from ..database import SessionLocal
from ..model import User
from sqlalchemy.exc import SQLAlchemyError
from fastapi import HTTPException
from app.schema.auth_schema import RegisterUser
from ..helper import Helper

db = SessionLocal()


class AuthService:
    @staticmethod
    def login(email: str, password: str) -> HTTPException | dict[str, str]:
        print(email, password)
        user = db.query(User).filter_by(email=email).first()
        if user is None or not Helper.match_hash_text(str(user.password), password):
            raise HTTPException(status_code=400, detail="Email or password wrong!")
        access_token = Helper.generate_access_token({"user_id": user.id})
        refresh_token = Helper.generate_access_token({"user_id": user.id})
        return {"access_token": access_token, "refresh_token": refresh_token}

    @staticmethod
    def register(payload: RegisterUser):
        try:
            user = User(
                email=payload.email,
                password=Helper.generate_hash_password(payload.password),
                role_id=1,
            )
            db.add(user)
            db.commit()
            return user
        except SQLAlchemyError as e:
            db.rollback()
            raise HTTPException(status_code=500, detail=str(e))

    @staticmethod
    def user_information(id: int):
        user = db.query(User).filter_by(id=id).first()
        if user:
            return user
        else:
            return False

File: node-server/src/services/auth-service.ts
import client from '../database';
import { BadRequestError } from '../exceptions/bad-request-exception';
import { Exception } from '../exceptions/exception';
import { InternalServerError } from '../exceptions/internal-server-exception';
import { NotFoundError } from '../exceptions/not-found-exception';
import {
    findByEmailQuery,
    loginQuery,
    registerQuery,
    updateUserPassword,
    verifyUserInformationQuery,
    verifyUserQuery,
} from '../queries/auth-queries';
import { Helper } from '../util/helper';

export class AuthService {
    private helper: Helper;

    constructor() {
        this.helper = new Helper();
    }

    public async login(email: string, password: string) {
        const hashPassword = this.helper.hashPassword(password);
        const result = await client.query(loginQuery, [email, hashPassword]);
        if (result.rows.length === 0) {
            throw new NotFoundError('User not found!');
        }

        const user = result.rows[0];
        const payload = {
            id: user.id,
            email: user.email,
        };
        const accessToken = this.helper.generateAccessToken(payload);
        const refreshToken = this.helper.generateRefreshToken(payload);
        return { access_token: accessToken, refresh_token: refreshToken };
    }

    public async register(payload: { email: string; password: string }) {
        const hashPassword = this.helper.hashPassword(payload.password);
        try {
            await client.query('BEGIN');
            const isEmailExist = await client.query(findByEmailQuery, [payload.email]);
            if (isEmailExist.rows.length) {
                throw new BadRequestError('Email already exist!');
            } else {
                const newUser = await client.query(registerQuery, [payload.email, hashPassword]);
                await client.query('COMMIT');
                return newUser;
            }
        } catch (error) {
            await client.query('ROLLBACK');
            if (error instanceof Exception) {
                throw error;
            } else {
                throw new InternalServerError('Internal Server Error!');
            }
        }
    }

    public async verify(token: string) {
        try {
            const decodedToken: any = this.helper.decodeToken(token);
            const result = await client.query(verifyUserQuery, [decodedToken.id]);
            if (result.rows.length === 0) {
                throw new NotFoundError('User not found!');
            }
            return result.rows[0];
        } catch (error) {
            if (error instanceof Exception) {
                throw error;
            } else {
                throw new InternalServerError('Internal Server Error!');
            }
        }
    }

    public async updatePassword(token: string, currentPassword: string, newPassoword: string) {
        try {
            const decodedToken: any = this.helper.decodeToken(token);
            const result = await client.query(verifyUserInformationQuery, [decodedToken.id]);
            if (result.rows.length === 0) {
                throw new NotFoundError('User not found!');
            }
            const user = result.rows[0];
            const isPasswordValid = this.helper.comparePassword(currentPassword, user.password);

            if (!isPasswordValid) {
                throw new BadRequestError('Current password is incorrect!');
            }

            const hashedNewPassword = this.helper.hashPassword(newPassoword);
            await client.query('BEGIN');
            await client.query(updateUserPassword, [hashedNewPassword, user.id]);
            await client.query('COMMIT');
            return { message: 'Password updated successfully!' };
        } catch (error) {
            await client.query('ROLLBACK');
            if (error instanceof Exception) {
                throw error;
            } else {
                throw new InternalServerError('Internal Server Error!');
            }
        }
    }
}
